From: olh@suse.de
Subject: fix memory corruption 

Bug 50213 - LTC12695

The fixes for 10gigi are: Redundant Write-Back in Receive Descriptor Ring
Caused Memory corruption. There are two fixes in the driver:
(1) Should leave 3 descriptors between the head and tail pointer of the ring.
(2) alloc rx has fix to wait one descriptor in case of double writeback.

--- ./drivers/net/ixgb/ixgb_main.c	2005-01-24 13:50:44.000000000 -0800
+++ ./drivers/net/ixgb/ixgb_main.c	2005-01-24 14:04:09.090895080 -0800
@@ -1969,8 +1969,8 @@ ixgb_alloc_rx_buffers(struct ixgb_adapte
 
 	num_group_tail_writes = IXGB_RX_BUFFER_WRITE;
 
-	/* leave one descriptor unused */
-	while(--cleancount > 0) {
+	/* leave three descriptors unused */
+	while(--cleancount > 2) {
 		rx_desc = IXGB_RX_DESC(*rx_ring, i);
 
 		skb = dev_alloc_skb(adapter->rx_buffer_len + NET_IP_ALIGN);
@@ -1997,6 +1997,10 @@ ixgb_alloc_rx_buffers(struct ixgb_adapte
 			               PCI_DMA_FROMDEVICE);
 
 		rx_desc->buff_addr = cpu_to_le64(buffer_info->dma);
+		/* guarantee DD bit not set now before h/w gets descriptor
+		 * this is the rest of the workaround for h/w double 
+		 * writeback. */
+		rx_desc->status = 0;
 
 		if((i & ~(num_group_tail_writes- 1)) == i) {
 			/* Force memory writes to complete before letting h/w
